File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solverGenerator/FastSolver/DLX1/DLXSolver.java
package de.sudoq.model.solverGenerator.FastSolver.DLX1;

import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import de.sudoq.model.solverGenerator.FastSolver.FastSolver;
import de.sudoq.model.sudoku.Field;
import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.PositionMap;
import de.sudoq.model.sudoku.Sudoku;

public class DLXSolver implements FastSolver {

    private boolean calculationDone;

    private List<int[][]> solutions;

    private AbstractSudokuSolver solver;
    private int[][] array;

    public DLXSolver(Sudoku s){

        switch (s.getSudokuType().getEnumType()){
            case standard16x16:
                solver = new Sudoku16DLX(); break;
            case samurai:
                solver = new DLXSudokuSamurai(); break;
            case Xsudoku:
                solver = new DLXSudokuX(); break;
            case standard9x9:
                solver = new SudokuDLX(); break;

            default:
                throw new IllegalArgumentException("only 16x16 are accepted at the moment!!!");
        }

        array = sudoku2Array(s);

        // SudokuDLX solver = new SudokuDLX();
       // solver.solve(hardest);

    }

    private static int[][] sudoku2Array(Sudoku s){
        Position dim = s.getSudokuType().getSize();
        int[][] sarray = new int[dim.getY()][dim.getX()];
        for (int r = 0; r < sarray.length; r++)
            for (int c = 0; c < sarray[0].length; c++){
                Field f = s.getField(Position.get(c,r));
                sarray[r][c] = f==null ? -1 //if pos doesn't exist e.g. (9,0) in SamuraiSudoku
                                       : f.isSolved() ? f.getCurrentValue() + 1
                                                      : 0;
            }
        return sarray;
    }


    @Override
    public boolean hasSolution() {
        if (!calculationDone){
            solver.solve(array);
            solutions = solver.getSolutions();
            calculationDone=true;
        }
        return solutions.size() > 0;
    }

    @Override
    public boolean isAmbiguous() {
        if (!calculationDone){
            /*
            * If the sudoku only has one solution, we have to go through the whole search space
            * to rule out further solutions.
            * as this takes a long time we rather want to stop after x minutes and treat it as if there are no further solution
            *
            * */
            ExecutorService executor = Executors.newSingleThreadExecutor();
            Collection c = Collections.singletonList(new AmbiguousTask());
            try {
                executor.invokeAll(c , 5, TimeUnit.MINUTES); // Timeout of 5 minutes.
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            executor.shutdown();
        }


        return solutions != null && solutions.size() > 1;
    }



    private class AmbiguousTask implements Callable{
        @Override
        public Object call() throws Exception {
            solver.solve(array);
            solutions = solver.getSolutions();
            calculationDone=true;
            return null;
        }
    }




    @Override
    public Position getAmbiguousPos() {
        int[][] first = solutions.get(0);
        int[][] second= solutions.get(1);

        for (int r = 0; r < first.length; r++) {
            for (int c = 0; c < first[0].length; c++) {
                if (first[r][c] != second[r][c])
                    return Position.get(c,r);
            }
        }
        throw new IllegalStateException();
    }

    @Override
    public PositionMap<Integer> getSolutions(){
        int[][] solution = solutions.get(0);
        PositionMap<Integer> pm = new PositionMap<>(Position.get(solution[0].length,solution.length));

        for (int r = 0; r < solution.length; r++)
            for (int c = 0; c < solution[0].length; c++){
                if (solution[r][c] != -1)
                    pm.put(Position.get(c,r), solution[r][c]-1);
            }
        return pm;
    }

    @Deprecated
    public List<int[][]> getBothSolutionsForDebugPurposes(){
        return solutions;
    }

}

File: sudoq-app/sudoqmodel/src/main/kotlin/de/sudoq/model/solverGenerator/FastSolver/DLX1/DLXSolver.kt
package de.sudoq.model.solverGenerator.FastSolver.DLX1

import de.sudoq.model.solverGenerator.FastSolver.FastSolver
import de.sudoq.model.sudoku.Position
import de.sudoq.model.sudoku.PositionMap
import de.sudoq.model.sudoku.Sudoku
import de.sudoq.model.sudoku.sudokuTypes.SudokuTypes
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

open class DLXSolver(s: Sudoku) : FastSolver {
    private var calculationDone = false

    private var solutions: MutableList<Array<IntArray>>? = null

    @get:Deprecated("")
    var bothSolutionsForDebugPurposes: MutableList<Array<IntArray>>? = solutions
        private set

    private val solver: AbstractSudokuSolver
    private val array: Array<IntArray?>

    init {
        when (s.sudokuType.enumType) {
            SudokuTypes.standard16x16 -> solver = Sudoku16DLX()
            SudokuTypes.samurai -> solver = DLXSudokuSamurai()
            SudokuTypes.Xsudoku -> solver = DLXSudokuX()
            SudokuTypes.standard9x9 -> solver = SudokuDLX()
            else -> throw IllegalArgumentException("only 16x16 are accepted at the moment!!!")
        }

        array = sudoku2Array(s)

        // SudokuDLX solver = new SudokuDLX();
        // solver.solve(hardest);
    }

    override fun hasSolution(): Boolean {
        if (!calculationDone) {
            solver.solve(array)
            this.bothSolutionsForDebugPurposes = solver.getSolutions()
            calculationDone = true
        }
        return bothSolutionsForDebugPurposes!!.isNotEmpty()
    }

    override fun isAmbiguous(): Boolean {
        if (!calculationDone) {
            /*
             * If the sudoku only has one solution, we have to go through the whole search space
             * to rule out further solutions.
             * as this takes a long time we rather want to stop after x minutes and treat it as if there are no further solution
             *
             * */
            val executor = Executors.newSingleThreadExecutor()
            val c: Collection<AmbiguousTask> = listOf(AmbiguousTask())
            try {
                executor.invokeAll(c, 5, TimeUnit.MINUTES) // Timeout of 5 minutes.
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            executor.shutdown()
        }


        return this.bothSolutionsForDebugPurposes != null && bothSolutionsForDebugPurposes!!.size > 1
    }

    override fun getAmbiguousPos(): Position {
        val first = bothSolutionsForDebugPurposes!![0]
        val second = bothSolutionsForDebugPurposes!![1]

        for (r in first.indices) {
            for (c in first[0].indices) {
                if (first[r][c] != second[r][c]) return Position[c, r]
            }
        }
        throw IllegalStateException()
    }

    override fun getSolutions(): PositionMap<Int> {
        val solution = bothSolutionsForDebugPurposes!![0]
        val pm: PositionMap<Int> = PositionMap(Position[solution[0].size, solution.size])

        for (r in solution.indices) for (c in solution[0].indices) {
            if (solution[r][c] != -1) pm.put(Position[c, r], solution[r][c] - 1)
        }
        return pm
    }

    private inner class AmbiguousTask : Callable<Any?> {
        @Throws(Exception::class)
        override fun call(): Any? {
            solver.solve(array)
            solutions = solver.getSolutions()
            calculationDone = true
            return null
        }
    }

    companion object {
        private fun sudoku2Array(s: Sudoku): Array<IntArray?> {
            val dim = s.sudokuType.size
            val sarray = Array<IntArray?>(dim.y) { IntArray(dim.x) }
            for (r in sarray.indices) for (c in sarray[0]!!.indices) {
                val f = s.getCellNullable(Position[c, r])
                sarray[r]!![c] = if (f == null)
                    -1 //if pos doesn't exist e.g. (9,0) in SamuraiSudoku
                else
                    if (f.isSolved)
                        f.currentValue + 1
                    else
                        0
            }
            return sarray
        }
    }
}

